src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,241 @@
+        }
+
+        if (productDTO.getStock() < quantity) {
+            throw new InvalidInputDataException("You can not remove more than: " + productDTO.getStock() + " "
+                    + productName + " Products from your shopping cart!");
+        }
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
-            throw new EntityNotFoundException("UserName: " + email + " not found in the Database!");
+            throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
-
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
 
-
-
-        if (productsInCart == 0) {
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
+        Cart cart;
+        if (cartRepository.findByUserId(user.getId()) != null) {
+            cart = cartRepository.findByUserId(user.getId());
+        } else {
+            throw new EntityNotFoundException("Cart not found in the Database!");
         }
 
-        double productTotalRemove = 0;
+        List<CartEntry> cartEntries;
+        Integer initialQuantity = 0;
 
-        List<Product> products = productList.stream().filter(p -> p.getProductName().equals(productName)).collect(Collectors.toList());
-        productList.removeAll(products);
+        cartEntries = cart.getEntries();
+        CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
 
-        cart.setTotal(cart.getTotal() - (products.size() * product.getProductPrice()));
-
-        for (int i = 0; i < quantity; i++) {
-            productList.add(product);
-            productTotalRemove = product.getProductPrice() + productTotalRemove;
+        if (cartEntry == null) {
+            throw new EntityNotFoundException("CartEntry not found in the Database!");
+        } else {
+            initialQuantity = cartEntry.getQuantity();
+            cartEntry.setQuantity(quantity);
+            cartEntry.setCart(cart);
         }
-        cart.setTotal(cart.getTotal() + productTotalRemove);
 
-        cart.setUser(user);
+        cart.setEntries(cartEntries);
 
-        cart.setProductList(productList);
+        double total = cart.getTotal() - (initialQuantity * productDTO.getPrice()) + (quantity * productDTO.getPrice());
 
-        cartRepository.save(cart);
-
-        return CartMapper.cartEntityToDto(cart);
+        cart.setTotal(total);
 
+        cartRepository.save(cart);
```

Comment:
you could have used the cart returned after the save and not to an additional get cart by user id

Reply:
Done! 🔝